[ZIP 0] Privacy Implications section and other updates

aquietinvestor opened the pull request:
- Require that a submitted ZIP draft MUST (rather than SHOULD) start in Draft status when published.
- Add references to "Reviewing a ZIP" from "ZIP Editor Responsibilities".
- Require newly published or substantively updated ZIP drafts to be announced on the Zcash Community Forum always (rather than "in some cases"), and also on the `#zips` channel of the R&D Discord. Give motivation and say what the announcements contain.
- In "Reviewing a ZIP":
  - Content review applies to updates as well as new ZIP drafts.
  - Describe what content review entails, including potentially requesting "input from community security or privacy experts".
  - Move the statement that "it is not the primary responsibility of the ZIP Editors to review proposals for security, correctness, or implementability" to this section.
  - Say that format review should generally occur after content review feedback has been addressed.
- Add "it fails to adequately consider privacy implications" as a potential reason for rejecting publication.
- In "ZIP format and structure":
  - Describe the Abstract section in more detail.
  - Describe a Privacy Implications section that "SHOULD be present if the ZIP makes trade-offs or assumptions that could impact the privacy of Zcash users  or protocol participants".
  - Describe the Requirements section.
  - Delete the suggestion to include a "Security and privacy considerations" section and the reference to RFC 3552.
- In "Specification of Status Workflow", require "a ZIP that has significant security or privacy implications" to have undergone "independent security review from qualified security of privacy experts" before it transitions to Released, or when an update "might invalidate the conclusions of previous security or privacy review, or leave significant security or privacy implications unaddressed".
- Use more consistent terminology: "ZIP draft" rather than "draft ZIP".
- Includes merge of #1004.

daira commented on zips/zip-0000.rst:
```diff
@@ -349,15 +351,18 @@ of the following reasons:
 
-Note that it is not the primary responsibility of the ZIP Editors to
-review proposals for security, correctness, or implementability.
+Note: While it is not the primary responsibility of the ZIP Editors to perform in-depth reviews of a proposal’s security, correctness, or implementability, they MUST verify that the proposal includes a Security and Privacy Considerations section that identifies relevant risks, trust assumptions, and privacy trade-offs.
```
```suggestion
Note: While it is not the primary responsibility of the ZIP Editors to perform in-depth reviews of a proposal’s security, correctness, or implementability, they MUST verify that the proposal identifies relevant risks, trust assumptions, and privacy trade-offs.
```

nuttycom commented on zips/zip-0000.rst:
```diff
@@ -476,10 +480,7 @@ Each ZIP SHOULD have the following parts:
-  and privacy considerations should be explicitly described, particularly
-  if the ZIP makes explicit trade-offs or assumptions. For guidance on
-  this section consider RFC 3552 [#RFC3552]_ as a starting point.
+* Security and privacy considerations — This section is mandatory for all ZIPs. Security and privacy considerations MUST be explicitly described, particularly if the ZIP makes trade-offs or assumptions that could impact the privacy, integrity, or safety of Zcash users. This includes, but is not limited to, trust assumptions introduced or modified by the proposal; privacy trade-offs or changes to  ...
```
In ZIP Sync, we discussed this section, and collectively agreed that the section is somewhat misplaced. The ZIP Editors propose the removal of this section, to be replaced by a `Privacy Implications` section to immediately follow the `Motivation` section. This `Privacy Implications` section is intended to contain an analysis of the privacy implications of the proposal and any privacy tradeoffs implied, and may contain forward references to the specification and associated rationale sections below.

There are two reasons for this change:
* The privacy implications of a particular proposed change are of paramount importance, and are of interest to general readers who may not read the detail of the specification. As such, this high level information should be presented at the top of the document.
* Security in particular is not something that can really be analyzed separately from the individual parts of the specification, so it doesn't make sense to have a _separate_ security section.

The ZIP Editors however also disagree that this section should be mandatory for all ZIPs, as a consequence of lessons that have been learned from the RFC process. Mandatory sections often get filled with boilerplate that does not serve to inform. It is probably sufficient to have a sentence in the abstract that says "This proposal has no implications for user privacy" if that is the case.

aquietinvestor replied:
I agree with replacing 'Security and Privacy Considerations' section with a 'Privacy Implications' section. Having the section at the top of the document will draw more attention to it since, as you say, most people don't read the details of the specification. 

I understand the concern about mandatory sections potentially leading to boilerplate responses, but I still believe there's value in making the 'Privacy Implications' section required. If a proposal has no implications for user privacy or security, then a simple statement like “This proposal has no implications for user privacy or security” is sufficient. The main point is that requiring the section prompts authors to explicitly consider and address these questions. If it’s optional, it’s much more likely to be overlooked. That’s already been the case historically. Requiring the section helps ensure that authors consider and document potential implications.

If the section remains optional, then the responsibility I added to the ZIP Editor role to "Ensure that the ZIP considers relevant security and privacy implications and adequately addresses and documents relevant risks, trust assumptions, and privacy trade-offs." becomes even more important. Making the section mandatory helps shift some of that burden onto ZIP Authors.

aquietinvestor commented:
First off, I agree with replacing the `Privacy and Security Considerations` section with a `Privacy Implications` section that immediately follows the `Motivation` section. I agree that this section should be presented at the top of the document because, as you say, general readers may not read the detail of the specification. 

I understand the concern about mandatory sections potentially leading to boilerplate responses, but I still believe there's value in making the `Privacy Implications` section required. If a proposal has no implications for user privacy, then a simple statement like “This proposal has no implications for user privacy” is sufficient. The main point is that requiring the section prompts authors to explicitly consider and address these questions, even if only to state that they’re not relevant.

The reason I think it should be required is if it remains optional, then it’s much more likely to be overlooked. That’s already been the case with most of the ZIPs historically. Requiring the section helps ensure that authors at least pause to consider the potential implications, even if the outcome is just a one-line statement that there are none. 

If the section remains optional, then the responsibility I added to the ZIP Editor role `to ensure that the ZIP considers relevant...privacy implications and documents key risks, trust assumptions, and trade-offs` becomes even more important. Making the section mandatory helps shift more of that burden onto ZIP authors. 

With regard a security section, I agree with your logic that it "is not something that can really be analyzed separately from the individual parts of the specification, so it doesn’t make sense to have a separate section.” I think this is okay so long as the following requirements I added are reflected in ZIP 0: (1) the ZIP Editors are required to ensure that the ZIP considers relevant security and privacy implications and documents key risks, assumption, and trade offs and (2) ZIP Authors are responsible for ensuring that the design and implementation undergo an independent security review.

defuse commented:
+1 to making the privacy implications section mandatory. If I'm auditing a ZIP and the section is there with a simple `This proposal has no implications for user privacy.` and then I find an impact on privacy, I've unambiguously found an error in the ZIP, whereas that could be ambiguous otherwise. A sentence in the abstract would play the same role but I worry it would be easier for both ZIP authors and auditors to overlook.

@nuttycom wrote:

> Security in particular is not something that can really be analyzed separately from the individual parts of the specification, so it doesn't make sense to have a separate security section.

While this is true, it's still important to summarize security implications at a high level. For example, suppose a ZIP were introduced that weakened Zcash's quantum-resistance-when-addresses-are-kept-secret property as a side-effect of implementing some protocol feature. That ZIP should clearly indicate that the security property is being changed without requiring the reader to dig in to the technical sections.

Ideally, there would be user-facing threat models for all Zcash components, similar to the wallet app threat model, and ZIPs would come with diffs to those threat models.

nuttycom commented on zips/zip-0000.rst:
```diff
@@ -256,13 +256,16 @@ For proposals to revise an existing ZIP, the ZIP Editors SHOULD:
-  for it to be sent to the Zcash Community Forum.
-* Ensure that motivation and backward compatibility have been
-  addressed, if applicable.
-* Ensure that the ZIP considers relevant security and privacy implications and adequately addresses and documents relevant risks, trust assumptions, and privacy trade-offs.
```
This has been moved to & rewrded in the `Reviewing a ZIP` section.

nuttycom reviewed (approved):
ACK b2e9fc5d61fbaa398b1bae1d6d7a1305ae57f52c

str4d commented:
Pushed ebe16be9cdf7803ecfad8d555e8b323c3abc0f91 making the changes discussed and worked on in the ZIP Sync meeting yesterday. These fell into several general categories:
- Replacing newly proposed text with rewording and clarification of already-existing language that already covered the relevant semantics.
- Moving both existing and newly proposed text to the correct sections and integrating it therein (some of the proposed text was contextually invalid or made no sense in the places it had been inserted).
- Removed duplication, in favour of fleshing out the "Reviewing a ZIP" section.
  - This is where the reminders on what to keep in mind during review are most relevant and useful.

During discussion, additional context from Sam led us into a discussion topic where we concluded that a top-level (as in a ones-series ZIP 000*) living ZIP focused on Zcash's "privacy story" would be a very useful addition, and would be a great place to expand on the kinds of user behaviours that Zcash is intended to protect. Future ZIPs could then make updates to this ZIP in the same way they do to e.g. ZIP 32.

(Also pushed a commit to clean up some existing issues with inconsistent usage of "ZIP draft".)

daira commented on zips/zip-0000.rst:
```diff
@@ -232,20 +232,20 @@ For each new ZIP that comes in, the ZIP Editors SHOULD:
-  for it to be sent to the Zcash Community Forum.
-* Ensure that motivation and backward compatibility have been
-  addressed, if applicable.
+* Ensure that a PR has been created for the ZIP draft in ZIPs git
```
```suggestion
* Ensure that a PR has been created for the ZIP draft in the ZIPs git
```

daira commented on zips/zip-0000.rst:
```diff
@@ -356,8 +380,6 @@ proposal or update, they MUST give an explanation of which of the
 
-Note that it is not the primary responsibility of the ZIP Editors to
-review proposals for security, correctness, or implementability.
 
```
```suggestion
```
Non-blocking nit.

daira commented on zips/zip-0000.rst:
```diff
@@ -410,13 +432,50 @@ Each ZIP SHOULD have the following parts:
+  trade-offs or assumptions that could impact the privacy of Zcash users
+  or protocol participants. Since this section is not part of the
+  Specification, it MUST NOT directly include or duplicate conformance
+  requirements on implementations. Instead it contains a high-level overview
+  of the privacy implications that may need to be considered for deployment
+  of the ZIP.
```
```suggestion
  requirements on implementations or processes. Instead it contains a
  high-level overview of the privacy implications that may need to be
  considered for deployment of the ZIP.
```

daira commented on zips/zip-0000.rst:
```diff
@@ -410,13 +432,50 @@ Each ZIP SHOULD have the following parts:
+  SHOULD assess the implications for user privacy in that context.
+
+* Requirements — This section describes high-level goals that the
+  Specification MUST ensure. It MUST NOT include conformance requirements,
+  so that it is possible to perform a consistency check that what is
+  specified meets the Requirements.
```
```suggestion
  Specification MUST ensure. It MUST NOT directly include conformance
  requirements, so that it is possible to perform a consistency check that
  what is specified meets the Requirements.
```

daira commented on zips/zip-0000.rst:
```diff
@@ -693,6 +747,11 @@ will specify a network upgrade in which the consensus change is to activate.
+A ZIP that has significant security or privacy implications MUST NOT
+be changed from a non-Released status to a Released status without the
+design having undergone independent security review from qualified
+security or privacy experts.
```
This potentially leaves a loophole where a change is made after a ZIP is Released that is substantive enough to require another independent security review. Typically I think we've handled that in the past by including all such changes made in a particular upgrade in the upgrade's design review, but the ZIP process should require that, I think.

These aren't sufficient to catch all such cases:
> * it disregards compatibility with the existing Zcash blockchain or ecosystem;
> * in the case of a Proposed, Active, Implemented, or Final ZIP, the update makes a substantive change to which there is significant community opposition;
> * an update to an existing ZIP extends or changes its scope to an extent that would be better handled as a separate ZIP;

daira reviewed (approved):
utACK modulo comments. This comment is blocking: https://github.com/zcash/zips/pull/1000/files/ebe16be9cdf7803ecfad8d555e8b323c3abc0f91#r2056866313

daira commented on zips/zip-0000.rst:
```diff
@@ -693,6 +746,14 @@ will specify a network upgrade in which the consensus change is to activate.
+A ZIP that has significant security or privacy implications MUST NOT
+be changed from a non-Released status to a Released status without the
+design having undergone independent security review from qualified
+security of privacy experts. Similarly, changes to a Released ZIP that might
```
```suggestion
security or privacy experts. Similarly, changes to a Released ZIP that might
```

daira commented:
> I understand the concern about mandatory sections potentially leading to boilerplate responses, but I still believe there's value in making the Privacy Implications section required. If a proposal has no implications for user privacy, then a simple statement like “This proposal has no implications for user privacy” is sufficient. The main point is that requiring the section prompts authors to explicitly consider and address these questions, even if only to state that they’re not relevant.

This policy only ends up encouraging vacuous, unhelpful, and often misleading boilerplate such as ["Security issues are discussed throughout this document"](https://datatracker.ietf.org/doc/html/rfc6331#section-2) found in some RFCs. If the ZIP Editors are following ZIP 0, they will call out the lack of a substantive Privacy Implications section, or omissions in that section, whenever it should be present. The current wording in the PR is:

> * Privacy Implications — This section SHOULD be present if the ZIP makes trade-offs or assumptions that could impact the privacy of Zcash users or protocol participants. [...]

SHOULD is the correct force here IMHO:

> [3](https://www.rfc-editor.org/rfc/rfc2119.html#section-3). SHOULD   This word, or the adjective "RECOMMENDED", mean that there may exist valid reasons in particular circumstances to ignore a particular item, but the full implications must be understood and carefully weighed before choosing a different course.

nuttycom reviewed (approved):
ACK 50b50e2ffc1bfc6cd31bb1da490ac2da3792605d
